File: apps/tinyolly/common/trace_lint.py
# BSD 3-Clause License
#
# Copyright (c) 2025, Infrastructure Architects, LLC
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
# 1. Redistributions of source code must retain the above copyright notice,
#    this list of conditions and the following disclaimer.
#
# 2. Redistributions in binary form must reproduce the above copyright notice,
#    this list of conditions and the following disclaimer in the documentation
#    and/or other materials provided with the distribution.
#
# 3. Neither the name of the copyright holder nor the names of its
#    contributors may be used to endorse or promote products derived from this
#    software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

"""
Trace Lint Module - Analyzes trace structures to identify unique flows and provide feedback.

This module computes trace flow signatures by hashing structural elements while ignoring
identifiers like IPs and user agents. It provides lint findings such as:
- Spans that should use auto-instrumentation
- Semantic convention violations
- Field naming issues
"""

import hashlib
import json
import re
from typing import Any

from .otlp_utils import get_attr_value, parse_attributes

# Semantic convention attributes (OpenTelemetry standard)
# https://opentelemetry.io/docs/specs/semconv/
SEMCONV_ATTRIBUTES = {
    "http.method",
    "http.request.method",
    "http.status_code",
    "http.response.status_code",
    "http.route",
    "http.target",
    "http.url",
    "http.scheme",
    "http.host",
    "http.server_name",
    "url.path",
    "url.full",
    "url.scheme",
    "net.host.name",
    "net.host.port",
    "net.peer.name",
    "net.peer.port",
    "db.system",
    "db.name",
    "db.statement",
    "db.operation",
    "messaging.system",
    "messaging.destination",
    "messaging.operation",
    "rpc.system",
    "rpc.service",
    "rpc.method",
    "error.type",
    "exception.type",
    "exception.message",
}

# Attributes that are identifiers (should be ignored for flow hashing)
IDENTIFIER_ATTRIBUTES = {
    "net.peer.ip",
    "net.host.ip",
    "client.address",
    "server.address",
    "http.client_ip",
    "user_agent.original",
    "http.user_agent",
    "enduser.id",
    "session.id",
    "trace.id",
    "span.id",
}

# Attributes that affect control flow (should be included in hash)
LOGIC_ATTRIBUTES = {
    "http.status_code",
    "http.response.status_code",
    "error.type",
    "exception.type",
    "exception.message",
    "db.operation",
    "messaging.operation",
    "rpc.method",
}


def _normalize_attribute_name(key: str) -> str:
    """Normalize attribute names to detect non-standard conventions."""
    return key.lower().replace("_", ".").replace("-", ".")


def _is_identifier_value(key: str, value: Any) -> bool:
    """Check if an attribute value is an identifier (IP, UUID, etc)."""
    if key in IDENTIFIER_ATTRIBUTES:
        return True

    if not isinstance(value, str):
        return False

    # IP address pattern
    if re.match(r"^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$", value):
        return True

    # UUID pattern
    if re.match(r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$", value.lower()):
        return True

    return False


def _extract_span_structure(span: dict[str, Any]) -> dict[str, Any]:
    """Extract structural elements from a span for flow hashing.

    Args:
        span: Span data dictionary

    Returns:
        Dictionary containing structural elements (excludes identifiers)
    """
    # Parse attributes if in list format
    attrs = span.get("attributes", {})
    if isinstance(attrs, list):
        attrs = parse_attributes(attrs)

    structure = {
        "name": span.get("name", ""),
        "kind": span.get("kind", 0),
        "serviceName": span.get("serviceName", "unknown"),
    }

    # Include logic-affecting attributes
    logic_attrs = {}
    for key, value in attrs.items():
        if not _is_identifier_value(key, value):
            # Include logic attributes and semantic conventions
            normalized_key = _normalize_attribute_name(key)
            if any(normalized_key.startswith(prefix) for prefix in ["http.", "db.", "rpc.", "messaging.", "error.", "exception."]):
                logic_attrs[key] = value

    if logic_attrs:
        structure["attributes"] = logic_attrs

    # Include status information
    status = span.get("status", {})
    if status:
        structure["status"] = {"code": status.get("code", 0)}

    return structure


def compute_trace_flow_hash(spans: list[dict[str, Any]]) -> str:
    """Compute a hash representing the trace flow structure.

    Args:
        spans: List of span dictionaries

    Returns:
        Hash string representing the trace flow
    """
    # Sort spans by start time to get consistent ordering
    sorted_spans = sorted(spans, key=lambda s: s.get("startTimeUnixNano", s.get("start_time", 0)))

    # Extract structures
    structures = [_extract_span_structure(span) for span in sorted_spans]

    # Create deterministic JSON representation
    flow_json = json.dumps(structures, sort_keys=True)

    # Compute hash
    return hashlib.sha256(flow_json.encode()).hexdigest()[:16]


def _check_attribute_naming(span: dict[str, Any]) -> list[dict[str, Any]]:
    """Check for attribute naming issues.

    Args:
        span: Span data dictionary

    Returns:
        List of lint findings
    """
    findings = []
    attrs = span.get("attributes", {})
    if isinstance(attrs, list):
        attrs = parse_attributes(attrs)

    for key in attrs.keys():
        normalized = _normalize_attribute_name(key)

        # Check for underscore instead of dot
        if "_" in key and "." not in key:
            # Check if there's a semantic convention equivalent
            for semconv in SEMCONV_ATTRIBUTES:
                if semconv.replace(".", "_") == key or semconv.replace(".", "_").lower() == key.lower():
                    findings.append(
                        {
                            "severity": "warning",
                            "type": "naming",
                            "message": f"Attribute '{key}' should use semantic convention '{semconv}'",
                            "suggestion": f"Rename '{key}' to '{semconv}'",
                            "span_name": span.get("name", "unknown"),
                        }
                    )
                    break

    # Check for missing standard HTTP attributes on HTTP spans
    span_name = span.get("name", "").lower()
    if any(method in span_name for method in ["get", "post", "put", "delete", "patch", "head", "options"]):
        if "http.method" not in attrs and "http.request.method" not in attrs:
            findings.append(
                {
                    "severity": "info",
                    "type": "missing_attribute",
                    "message": f"HTTP span '{span.get('name')}' missing http.method attribute",
                    "suggestion": "Add http.method or http.request.method attribute for HTTP spans",
                    "span_name": span.get("name", "unknown"),
                }
            )

    return findings


def _check_auto_instrumentation(span: dict[str, Any]) -> list[dict[str, Any]]:
    """Check if span should be handled by auto-instrumentation.

    Args:
        span: Span data dictionary

    Returns:
        List of lint findings
    """
    findings = []
    span_name = span.get("name", "")
    attrs = span.get("attributes", {})
    if isinstance(attrs, list):
        attrs = parse_attributes(attrs)

    # Check for HTTP client/server spans
    http_method = get_attr_value(span, ["http.method", "http.request.method"])
    if http_method:
        kind = span.get("kind", 0)
        # Server spans (kind=2) and client spans (kind=3)
        if kind in [2, 3]:
            # Check if this looks like a manually instrumented span
            if not any(key.startswith("http.") for key in attrs.keys()):
                findings.append(
                    {
                        "severity": "info",
                        "type": "auto_instrumentation",
                        "message": f"HTTP span '{span_name}' may benefit from auto-instrumentation",
                        "suggestion": "Consider using OpenTelemetry auto-instrumentation for HTTP requests",
                        "span_name": span_name,
                    }
                )

    # Check for database spans
    db_system = get_attr_value(span, ["db.system"])
    if db_system or any(db in span_name.lower() for db in ["sql", "select", "insert", "update", "delete", "query", "database", "mongo", "redis", "postgres", "mysql"]):
        if not db_system:
            findings.append(
                {
                    "severity": "warning",
                    "type": "missing_attribute",
                    "message": f"Database span '{span_name}' missing db.system attribute",
                    "suggestion": "Add db.system attribute for database operations",
                    "span_name": span_name,
                }
            )

    return findings


def lint_trace(spans: list[dict[str, Any]]) -> dict[str, Any]:
    """Lint a trace and return findings.

    Args:
        spans: List of span dictionaries

    Returns:
        Dictionary containing flow hash and lint findings
    """
    flow_hash = compute_trace_flow_hash(spans)
    findings = []

    for span in spans:
        # Check attribute naming
        findings.extend(_check_attribute_naming(span))

        # Check auto-instrumentation opportunities
        findings.extend(_check_auto_instrumentation(span))

    # Deduplicate findings by message
    unique_findings = []
    seen_messages = set()
    for finding in findings:
        msg = finding["message"]
        if msg not in seen_messages:
            seen_messages.add(msg)
            unique_findings.append(finding)

    return {
        "flow_hash": flow_hash,
        "findings": unique_findings,
        "span_count": len(spans),
        "severity_counts": {
            "error": len([f for f in unique_findings if f["severity"] == "error"]),
            "warning": len([f for f in unique_findings if f["severity"] == "warning"]),
            "info": len([f for f in unique_findings if f["severity"] == "info"]),
        },
    }


def compute_flow_summary(spans: list[dict[str, Any]]) -> dict[str, Any]:
    """Compute a summary of the trace flow for display.

    Args:
        spans: List of span dictionaries

    Returns:
        Dictionary containing flow summary information
    """
    if not spans:
        return {}

    # Sort by start time
    sorted_spans = sorted(spans, key=lambda s: s.get("startTimeUnixNano", s.get("start_time", 0)))

    # Get root span (first span, typically)
    root_span = sorted_spans[0]

    # Extract key attributes
    attrs = root_span.get("attributes", {})
    if isinstance(attrs, list):
        attrs = parse_attributes(attrs)

    http_method = get_attr_value(root_span, ["http.method", "http.request.method"])
    http_route = get_attr_value(root_span, ["http.route", "http.target", "url.path"])
    status_code = get_attr_value(root_span, ["http.status_code", "http.response.status_code"])

    # Build service chain
    services = []
    for span in sorted_spans:
        service = span.get("serviceName", "unknown")
        if service not in services:
            services.append(service)

    return {
        "root_span_name": root_span.get("name", "unknown"),
        "root_service": root_span.get("serviceName", "unknown"),
        "span_count": len(spans),
        "service_chain": services,
        "http_method": http_method,
        "http_route": http_route,
        "status_code": status_code,
    }

File: apps/tinyolly-ui/src/modules/traceLint.js
/*
 * BSD 3-Clause License
 *
 * Copyright (c) 2025, Infrastructure Architects, LLC
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its
 *    contributors may be used to endorse or promote products derived from this
 *    software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/**
 * Trace Lint Module - Handles trace flow analysis and lint findings display
 */
import { renderEmptyState, renderErrorState, renderLoadingState, renderTableHeader } from './utils.js';

let currentFlowHash = null;

/**
 * Load and display trace flows
 */
export async function loadTraceFlows() {
    const container = document.getElementById('trace-flows-container');
    if (!container) {
        console.error('Trace flows container not found');
        return;
    }

    container.innerHTML = renderLoadingState('Loading trace flows...');

    try {
        const response = await fetch('/api/trace-flows?limit=50');
        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }

        const flows = await response.json();
        renderTraceFlows(flows);
    } catch (error) {
        console.error('Error loading trace flows:', error);
        container.innerHTML = renderErrorState('Error loading trace flows');
    }
}

/**
 * Render trace flows list
 */
function renderTraceFlows(flows) {
    const container = document.getElementById('trace-flows-container');

    if (!flows || flows.length === 0) {
        container.innerHTML = renderEmptyState('No trace flows yet. Send some traces to get started!');
        return;
    }

    const headerRow = renderTableHeader([
        { label: 'Flow Hash', flex: '0 0 140px' },
        { label: 'Root Operation', flex: '1' },
        { label: 'Service', flex: '0 0 120px' },
        { label: 'Method', flex: '0 0 70px' },
        { label: 'Route', flex: '1' },
        { label: 'Spans', flex: '0 0 60px', align: 'right' },
        { label: 'Traces', flex: '0 0 70px', align: 'right' },
        { label: 'Findings', flex: '0 0 80px', align: 'right' }
    ]);

    const flowsHtml = flows.map(flow => {
        const findingCount = flow.finding_count || 0;
        const severityCounts = flow.severity_counts || { error: 0, warning: 0, info: 0 };
        
        // Determine color based on severity
        let findingColor = 'var(--text-muted)';
        if (severityCounts.error > 0) {
            findingColor = '#ef4444'; // Red for errors
        } else if (severityCounts.warning > 0) {
            findingColor = '#f59e0b'; // Orange for warnings
        } else if (severityCounts.info > 0) {
            findingColor = '#3b82f6'; // Blue for info
        }

        const method = flow.http_method || '';
        const route = flow.http_route || flow.root_span_name || '-';
        const service = flow.root_service || 'unknown';
        const flowHashShort = flow.flow_hash.substring(0, 12);

        return `
            <div class="trace-flow-item data-table-row" data-flow-hash="${flow.flow_hash}">
                <div class="flow-hash text-mono text-muted" style="flex: 0 0 140px; font-size: 0.85em;">${flowHashShort}...</div>
                <div class="flow-name font-medium text-truncate" style="flex: 1;" title="${flow.root_span_name}">${flow.root_span_name}</div>
                <div class="flow-service text-truncate" style="flex: 0 0 120px;" title="${service}">${service}</div>
                <div class="flow-method text-primary font-bold text-truncate" style="flex: 0 0 70px;">${method}</div>
                <div class="flow-route text-truncate" style="flex: 1;" title="${route}">${route}</div>
                <div class="flow-spans text-muted" style="flex: 0 0 60px; text-align: right;">${flow.span_count}</div>
                <div class="flow-trace-count text-muted" style="flex: 0 0 70px; text-align: right;">${flow.trace_count}</div>
                <div class="flow-findings font-medium" style="flex: 0 0 80px; text-align: right; color: ${findingColor};">
                    ${findingCount > 0 ? findingCount : '-'}
                </div>
            </div>
        `;
    }).join('');

    container.innerHTML = `
        <div class="data-table">
            ${headerRow}
            ${flowsHtml}
        </div>
    `;

    // Add click handlers
    document.querySelectorAll('.trace-flow-item').forEach(item => {
        item.addEventListener('click', () => {
            const flowHash = item.dataset.flowHash;
            showFlowDetail(flowHash);
        });
    });
}

/**
 * Show detailed view of a trace flow
 */
async function showFlowDetail(flowHash) {
    currentFlowHash = flowHash;
    
    const listView = document.getElementById('trace-flows-list-view');
    const detailView = document.getElementById('trace-flow-detail-view');
    const detailContainer = document.getElementById('trace-flow-detail-container');

    if (!listView || !detailView || !detailContainer) {
        console.error('Flow detail view containers not found');
        return;
    }

    // Show detail view
    listView.style.display = 'none';
    detailView.style.display = 'block';

    detailContainer.innerHTML = renderLoadingState('Loading flow details...');

    try {
        const response = await fetch(`/api/trace-flows/${flowHash}`);
        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }

        const flowDetail = await response.json();
        renderFlowDetail(flowDetail);
    } catch (error) {
        console.error('Error loading flow detail:', error);
        detailContainer.innerHTML = renderErrorState('Error loading flow details');
    }
}

/**
 * Render detailed flow view
 */
function renderFlowDetail(flowDetail) {
    const container = document.getElementById('trace-flow-detail-container');
    const summary = flowDetail.summary || {};
    const lintResult = flowDetail.lint_result || {};
    const findings = lintResult.findings || [];
    const exampleTraces = flowDetail.example_traces || [];

    // Back button
    const backButton = `
        <div style="margin-bottom: 20px;">
            <button onclick="window.backToTraceFlows()" class="btn-secondary">
                ← Back to Trace Flows
            </button>
        </div>
    `;

    // Summary section
    const summarySection = `
        <div class="card" style="margin-bottom: 20px;">
            <h3 style="margin: 0 0 15px 0;">Flow Summary</h3>
            <div style="display: grid; grid-template-columns: repeat(auto-fit, minmax(200px, 1fr)); gap: 15px;">
                <div>
                    <div class="text-muted" style="font-size: 0.85em; margin-bottom: 4px;">Flow Hash</div>
                    <div class="text-mono" style="font-size: 0.9em;">${summary.flow_hash || ''}</div>
                </div>
                <div>
                    <div class="text-muted" style="font-size: 0.85em; margin-bottom: 4px;">Root Operation</div>
                    <div class="font-medium">${summary.root_span_name || 'unknown'}</div>
                </div>
                <div>
                    <div class="text-muted" style="font-size: 0.85em; margin-bottom: 4px;">Root Service</div>
                    <div>${summary.root_service || 'unknown'}</div>
                </div>
                <div>
                    <div class="text-muted" style="font-size: 0.85em; margin-bottom: 4px;">HTTP Method</div>
                    <div class="text-primary font-bold">${summary.http_method || '-'}</div>
                </div>
                <div>
                    <div class="text-muted" style="font-size: 0.85em; margin-bottom: 4px;">HTTP Route</div>
                    <div>${summary.http_route || '-'}</div>
                </div>
                <div>
                    <div class="text-muted" style="font-size: 0.85em; margin-bottom: 4px;">Span Count</div>
                    <div>${summary.span_count || 0}</div>
                </div>
                <div>
                    <div class="text-muted" style="font-size: 0.85em; margin-bottom: 4px;">Trace Count</div>
                    <div>${summary.trace_count || 0}</div>
                </div>
                <div>
                    <div class="text-muted" style="font-size: 0.85em; margin-bottom: 4px;">Status Code</div>
                    <div>${summary.status_code || '-'}</div>
                </div>
            </div>
            ${summary.service_chain && summary.service_chain.length > 0 ? `
                <div style="margin-top: 15px;">
                    <div class="text-muted" style="font-size: 0.85em; margin-bottom: 8px;">Service Chain</div>
                    <div style="display: flex; gap: 8px; flex-wrap: wrap;">
                        ${summary.service_chain.map(service => `
                            <span class="badge">${service}</span>
                        `).join('<span style="color: var(--text-muted);">→</span>')}
                    </div>
                </div>
            ` : ''}
        </div>
    `;

    // Lint findings section
    const severityCounts = lintResult.severity_counts || { error: 0, warning: 0, info: 0 };
    const findingsSection = `
        <div class="card" style="margin-bottom: 20px;">
            <h3 style="margin: 0 0 15px 0;">
                Lint Findings
                <span class="badge" style="background: ${severityCounts.error > 0 ? '#ef4444' : severityCounts.warning > 0 ? '#f59e0b' : '#3b82f6'}; color: white; margin-left: 10px;">
                    ${findings.length} total
                </span>
            </h3>
            ${findings.length === 0 ? `
                <div class="empty-state">
                    <div style="font-size: 2em; margin-bottom: 10px;">✓</div>
                    <div>No issues found. This trace flow follows best practices!</div>
                </div>
            ` : `
                <div style="display: flex; gap: 10px; margin-bottom: 15px;">
                    ${severityCounts.error > 0 ? `<span class="badge" style="background: #ef4444; color: white;">${severityCounts.error} errors</span>` : ''}
                    ${severityCounts.warning > 0 ? `<span class="badge" style="background: #f59e0b; color: white;">${severityCounts.warning} warnings</span>` : ''}
                    ${severityCounts.info > 0 ? `<span class="badge" style="background: #3b82f6; color: white;">${severityCounts.info} info</span>` : ''}
                </div>
                <div style="display: flex; flex-direction: column; gap: 12px;">
                    ${findings.map((finding, idx) => {
                        let severityColor = '#3b82f6'; // info
                        let severityIcon = 'ℹ️';
                        if (finding.severity === 'error') {
                            severityColor = '#ef4444';
                            severityIcon = '❌';
                        } else if (finding.severity === 'warning') {
                            severityColor = '#f59e0b';
                            severityIcon = '⚠️';
                        }

                        return `
                            <div class="finding-item" style="border-left: 3px solid ${severityColor}; padding: 12px; background: var(--card-bg); border-radius: 4px;">
                                <div style="display: flex; align-items: start; gap: 10px;">
                                    <div style="font-size: 1.2em; flex-shrink: 0;">${severityIcon}</div>
                                    <div style="flex: 1;">
                                        <div style="font-weight: 600; margin-bottom: 4px; color: ${severityColor}; text-transform: uppercase; font-size: 0.85em;">
                                            ${finding.severity} - ${finding.type}
                                        </div>
                                        <div style="margin-bottom: 8px;">${finding.message}</div>
                                        <div style="background: var(--code-bg); padding: 8px; border-radius: 4px; font-size: 0.9em; color: var(--text-muted);">
                                            <strong>Suggestion:</strong> ${finding.suggestion}
                                        </div>
                                        <div class="text-muted" style="font-size: 0.85em; margin-top: 6px;">
                                            Span: <span class="text-mono">${finding.span_name}</span>
                                        </div>
                                    </div>
                                </div>
                            </div>
                        `;
                    }).join('')}
                </div>
            `}
        </div>
    `;

    // Example traces section
    const examplesSection = `
        <div class="card">
            <h3 style="margin: 0 0 15px 0;">Example Traces</h3>
            ${exampleTraces.length === 0 ? `
                <div class="empty-state">No example traces available</div>
            ` : `
                <div style="display: flex; flex-direction: column; gap: 8px;">
                    ${exampleTraces.map(traceId => `
                        <div class="example-trace-item" style="display: flex; justify-content: space-between; align-items: center; padding: 10px; background: var(--code-bg); border-radius: 4px;">
                            <div class="text-mono" style="font-size: 0.9em;">${traceId}</div>
                            <button class="btn-secondary btn-sm" onclick="window.viewTrace('${traceId}')">
                                View Trace →
                            </button>
                        </div>
                    `).join('')}
                </div>
            `}
        </div>
    `;

    container.innerHTML = backButton + summarySection + findingsSection + examplesSection;
}

/**
 * Go back to trace flows list
 */
export function backToTraceFlows() {
    const listView = document.getElementById('trace-flows-list-view');
    const detailView = document.getElementById('trace-flow-detail-view');

    if (listView && detailView) {
        listView.style.display = 'block';
        detailView.style.display = 'none';
        currentFlowHash = null;
    }
}

/**
 * View a specific trace (switches to traces tab)
 */
export function viewTrace(traceId) {
    // Import switchTab dynamically to avoid circular dependency
    import('./tabs.js').then(({ switchTab }) => {
        // Switch to traces tab
        const tracesTab = document.querySelector('[data-tab="traces"]');
        if (tracesTab) {
            switchTab('traces', tracesTab);
            
            // Trigger trace detail view
            setTimeout(() => {
                const traceItem = document.querySelector(`[data-trace-id="${traceId}"]`);
                if (traceItem) {
                    traceItem.click();
                }
            }, 100);
        }
    });
}

// Export for global access
window.backToTraceFlows = backToTraceFlows;
window.viewTrace = viewTrace;
